### DISCUSSION TITLE
createModelReference

### DISCUSSION THREAD

**Discussion by cdietrich** [TypeFox/monaco-languageclient#721] [Q&A]
Hi, do know how `createModelReference` and what is actually sent to the language server.
something like

```
  const modelRef = await createModelReference(Uri.parse("dully.our.extension"), `package x entity Y {}`)
  modelRef.object.setLanguageId("our-language-id")
```

works

now we try to funnel addtional config files to the server

```
const modelRef2 = await createModelReference(Uri.parse("dully.json"), `{"a":"b"}`)
modelRef2.object.setLanguageId("json")
```
but how will this end up in the server.

do you have another idea how we can funnel config to the server?

**Comment by CGNonofr** [comment]
It's your responsability to keep client and server filesystem in sync.

Btw, I don't recommend using the second parameter of `createModelReference`, you better register your file in the virtual filesystem, or using a custom filesystem provider, then use `createModelReference`.

What I recommend is:
- You first create all the required files on the server filesystem
- You implement your custom FileSystemProvider, that read and write files on your server via the protocol of your choice
- You use `createModelReference` to open files that exists on your server

**Reply by cdietrich** [reply]
is there any example for that?
there is no server file system. all communication is done by client. 
also we dont have any side channel to the server.
its just created as a webworker. (langium)
and of course this communication channel is occupied by lsp

(why does it work with file extensions the server knows)

and do all on webworker creation also would not really work nicely and we would have to destroy each time the non primary files change (instead of just disposing the modelRefs)

**Reply by CGNonofr** [reply]
ok, what language client are you using? Doesn't it need to access the filesystem?

> and of course this communication channel is occupied by lsp

The LSP protocol is based on JSONRpc, and it can easily be extended though

> and do all on webworker creation also would not really work nicely and we would have to destroy each time the non primary files change (instead of just disposing the modelRefs)

Not sure to understand what you would have to destroy and why

**Reply by cdietrich** [reply]
i just use `monaco-editor-wrapper` (either direct or via react) and talk to langium language server.
all content is fed to it by react in production and hardcoded in my hello world example / the langium web example

if we switch the single file edited (as the wrapper does if you change the file that is currently edited)
we also close the compangion files and replace them with others.

using modelRef it works ootb.

extending the protocol is also tedious
as langium relys on TextDocuments magic and not actively manage everything (didOpen)

**Reply by CGNonofr** [reply]
I'm not really familiar with that language server and worker language servers, maybe @kaisalmen will be able to help you

**Reply by cdietrich** [reply]
https://github.com/cdietrich/newer-mew/blob/11e0cd018c13d4015c49e465975c5308a5af61b9/src/setupClassic.ts#L88

**Reply by kaisalmen** [reply]
Hi @cdietrich are you aware of this:
https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/main.ts#L46-L47

Two files are first registered with the fileSystemProvider and are then opened using the vscode api. This triggers an event via the LSP. With Pyright the problem is it does not care about it, but if you define the LS code as well/it is under your control you see `textDocument/didOpen` with the file content. Does that help?

**Reply by cdietrich** [reply]
before i used didOpen and you told modelRef is better. (as it also handles did close)

**Reply by cdietrich** [reply]
also. what i dont understand yet. how are the initial files sent to the python server

**Reply by kaisalmen** [reply]
I did? Could be. 😊 The whole file handling in a distributed context is not trivial and I learn while going forward as well. When different Language Servers handle things differently it makes it even more complicated. If the client is the driver of you app, the easiest way possibly forward I see is to intercept message on the language server side like done [here](https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/langium/statemachine/worker/statemachine-server-port.ts) and handle the file contents. 

> also. what i dont understand yet. how are the initial files sent to the python server

`vscode.workspace.openTextDocument` triggers those LSP messages in the end:
![image](https://github.com/user-attachments/assets/509d375c-b248-4714-a54d-186b70677f52)

A general file sync utility utilizing LSP could really help to overcome those problems (web worker or distributed node backend)

**Reply by cdietrich** [reply]
ok so we would have to extend the protocol.

**Reply by cdietrich** [reply]
so
```
const fileSystemProvider = new RegisteredFileSystemProvider(false);
    fileSystemProvider.registerFile(new RegisteredMemoryFile(helloPyUri, helloPyCode));
    fileSystemProvider.registerFile(new RegisteredMemoryFile(hello2PyUri, hello2PyCode));
    ``` is a client side thing.

**Reply by kaisalmen** [reply]
Not necessarily, but one side must be the "driver", either client or server.
@CGNonofr do you have more resources / docs regarding usage of file system providers in a worker context or general file system sync recipes with LSP?

**Reply by cdietrich** [reply]
the concept langium uses is "EmptyFileSystem" from languim. thus it purely relys on client side driver

**Reply by kaisalmen** [reply]
You could provide your own implementation. but the API is based on node (Quote: "The default implementation is based on the node.js `fs` API."). But the question is if that is needed.

**Reply by CGNonofr** [reply]
> vscode.workspace.openTextDocument triggers those LSP messages in the end:

FYI `vscode.workspace.openTextDocument` just call createModelReference internally, and dispose the reference after a timeout

> @CGNonofr do you have more resources / docs regarding usage of file system providers in a worker context or general file system sync recipes with LSP?

I can tell you how @CodinGame: we've extended the LSP to add 4 new methods `file/read`, `file/write`, `file/stats` and `file/readdir`. The server act as a LSP proxy, intercept all requests, handle those 4 requests, and forward everything else to the delegate language server.
On the client side:
- We add in the initialization request `initializationOptions` all the open file content in the client, the server proxy use it to initialize the filesystem then remove it from the request
- we use the 4 additional methods to implement a custom FileSystemProvider we registrer

Does it answer your question? I don't have much to say about workers

**Reply by kaisalmen** [reply]
> Does it answer your question? I don't have much to say about workers

Yes, thank you that helps. 

> I can tell you how https://github.com/CodinGame: we've extended the LS

Is that code open or closed source? I was thinking if we should have something here. We had a discussion about server tools in the past in some issue, but I don't remember where exactly.

**Reply by CGNonofr** [reply]
The client code is at https://github.com/CodinGame/monaco-languageclient-wrapper/blob/main/src/customRequests.ts / https://github.com/CodinGame/monaco-languageclient-wrapper/blob/b7711e9c3be026d4c1b495f91af384cc73627a35/src/extensions.ts#L144 / https://github.com/CodinGame/monaco-languageclient-wrapper/blob/b7711e9c3be026d4c1b495f91af384cc73627a35/src/infrastructure.ts#L121-L141

The server code is not published anywhere public

**Reply by kaisalmen** [reply]
@CGNonofr thank you. I think we should integrate this functionality here (addition to monaco-languageclient itself), at least client and maybe also something reusable for the server. We could demonstrate the application with pyright.

**Reply by CGNonofr** [reply]
I agree, that's a very common pain point
